### PR TITLE
Added some examples of possible validations for testbed.yaml file

### PR DESCRIPTION
**Pull request by SergiyBeno**

### CODE REVIEW

File: tests/testbed_validation_rules.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        server_ip = testbed.servers.server_alias.address
+        assert re.fullmatch(self.ip_pattern, server_ip)
+        return True
+
+    def __str__(self):
+        return "Check server ip format"
+
+
+class Rules:
+    def __init__(self, set_of_rules):
```
Comment (extsoft) [inline tests/testbed_validation_rules.py]: @SergiyBeno just use simple names and don't forget about type hints
```suggestion
    def __init__(self, rules: Sequence[Rule]):
```

File: tests/testbed_validation_rules.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        self._set_of_rules = set_of_rules
+
+    def validate(self, testbed: Testbed):
+        passed_rules, failed_rules = [], []
+        for rule_template in self._set_of_rules:
+            try:
+                rule = rule_template()
```
Comment (extsoft) [inline tests/testbed_validation_rules.py]: @SergiyBeno it's a bad idea to instantiate rules here. You have to pass instances of rules (instead of classes) to the constructor.

File: tests/testbed_validation_rules.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                result = rule.is_passed(testbed)
+            except Exception:
+                failed_rules.append(str(rule))
+                continue
+            passed_rules.append(str(rule)) if result else failed_rules.append(str(rule))
+        _log.info(f'Passed rules:\n"{passed_rules}"')
+        _log.warning(f'Failed rules:\n"{failed_rules}"')
```
Comment (extsoft) [inline tests/testbed_validation_rules.py]: @SergiyBeno the execution has to stop if there is at least one failed rule.

File: tests/testbed_validation_rules.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    def validate(self, testbed: Testbed):
+        passed_rules, failed_rules = [], []
+        for rule_template in self._set_of_rules:
+            try:
+                rule = rule_template()
+                result = rule.is_passed(testbed)
+            except Exception:
+                failed_rules.append(str(rule))
+                continue
+            passed_rules.append(str(rule)) if result else failed_rules.append(str(rule))
```
Comment (extsoft) [inline tests/testbed_validation_rules.py]: @SergiyBeno please move it to `try` block.

File: tests/testbed_validation_rules.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        return "Check devices ip format"
+
+
+class CheckSeleniumGridUrl(Rule):
+
+    url_pattern = r"^https?:\/\/.+:\d+$"
+
+    def is_passed(self, testbed: Testbed) -> bool:
+        grid_url = testbed.custom["selenium-grid"]
+        assert re.fullmatch(self.url_pattern, grid_url)
```
Comment (extsoft) [inline tests/testbed_validation_rules.py]: @SergiyBeno I think 
```suggestion
        return re.fullmatch(self.url_pattern, grid_url)
```
is more suitable.

And please check it for other rules also.

File: tests/testbed_validation_rules.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    def __str__(self):
+        return "Check user credentials"
+
+
+class CheckDevicesIp(Rule):
+    def is_passed(self, testbed: Testbed) -> bool:
+        for device in testbed:
+            device_ip = str(device.connections.main.ip)
+            assert re.fullmatch(self.ip_pattern, device_ip)
```
Comment (extsoft) [inline tests/testbed_validation_rules.py]: @SergiyBeno why do you use `assert` here? Looks like you need `return False`.

File: tests/testbed_validation_rules.py (hunk 5 above)
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: Yes, indeed - I'll replace.

File: tests/testbed_validation_rules.py (hunk 1 above)
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: Sure, my bad. Fixed.

File: tests/testbed_validation_rules.py (hunk 3 above)
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: Ok, roger that.

File: tests/testbed_validation_rules.py (hunk 4 above)
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: It is done.

File: tests/testbed_validation_rules.py (hunk 2 above)
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: Fixed.

Comment (extsoft) [reply tests/testbed_validation_rules.py]: @extsoft sorry, but it is not fixed. I still see `rule_template()`.

File: suite.py [hunk 7]
```diff
@@ -90,6 +101,7 @@ def tests_runner(test_suite: Tuple, instance: Any) -> List:  # type: ignore
 
 
 def main(runtime: EasypyRuntime) -> None:
+    Rules(set_of_rules=_testbed_rules).validate(testbed=runtime.testbed)
```
Comment (extsoft) [inline suite.py]: @SergiyBeno `set_of_rules` is not a valid name.

File: oct/tests/__init__.py [hunk 8]
```diff
@@ -16,4 +17,7 @@ def mandatory_aetest_arguments(testbed: Testbed, device_name: str) -> Dict[str,
 
 
 def run_testcase(testbed_file: str = "testbed.yaml", device_name: str = "vm") -> None:
-    aetest.main(**mandatory_aetest_arguments(loader.load(testbed_file), device_name))
+    from suite import _testbed_rules
+    testbed = loader.load(testbed_file)
+    Rules(set_of_rules=_testbed_rules).validate(testbed)
```
Comment (extsoft) [inline oct/tests/__init__.py]: @SergiyBeno `set_of_rules` is not a valid name

Comment (SergiyBeno) [reply oct/tests/__init__.py]: Fixed

File: suite.py (hunk 7 above)
Comment (SergiyBeno) [reply suite.py]: Fixed.

File: tests/testbed_validation_rules.py (hunk 2 above)
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: Probably I forgot to push the changes to remote. Fixed.

File: tests/testbed_validation_rules.py (hunk 6 above)
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: I can't use return statement since I assume that I have more than 1 device to check. In that case it will validate only first device. I can modify it to check only 1 device if necessary.

Comment (extsoft) [reply tests/testbed_validation_rules.py]: If a first IP will be incorrect, `assert` won't allow you to continue.

File: tests/testbed_validation_rules.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+    def __str__(self) -> str:
+        return "Check server ip format"
+
+
+class Rules:
+    def __init__(self, rules: Sequence[Rule]):
+        self._rules = rules
+
+    def validate(self, testbed: Testbed) -> Tuple[bool, Any]:
```
Comment (extsoft) [inline tests/testbed_validation_rules.py]: @SergiyBeno it's a bad idea to return a tuple. I think it's better to return a status which is going to tell if validation is passed or not and will be able to print a statistic (for instance which rules are failed).

File: suite.py [hunk 10]
```diff
@@ -90,11 +99,15 @@ def tests_runner(test_suite: Tuple, instance: Any) -> List:  # type: ignore
...
-        tests_runner(_destroy_test, runtime.testbed)
+    all_rules_passed, failing_rule = Rules(rules=_testbed_rules).validate(testbed=runtime.testbed)
+    if all_rules_passed:
+        if "passed" in tests_runner(_deployment_test, runtime.testbed):
+            time.sleep(60)
+            if "failed" not in tests_runner(_api_tests, runtime.testbed):
+                tests_runner(_web_tests, runtime.testbed)
+            tests_runner(_destroy_test, runtime.testbed)
+    else:
+        _log.error(msg=f'Tests execution aborted due to failing "{failing_rule}" validation.')
```
Comment (extsoft) [inline suite.py]: @SergiyBeno you have to fail execution if there is at least one error.

File: tests/testbed_validation_rules.py (hunk 6 above)
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: Fixed - now we are checking all ips.

File: suite.py (hunk 10 above)
Comment (SergiyBeno) [reply suite.py]: Fixed - replaced with AssertionError.

File: tests/testbed_validation_rules.py (hunk 9 above)
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: But it sounds like my current implementation - tuple consists of overall rules validations flag (bool) and explanation about which exactly rule has failed (str):
`all_rules_passed, failing_rule = Rules(rules=testbed_rules).validate(testbed=runtime.testbed)`

File: .pylintrc [hunk 11]
```diff
@@ -39,7 +39,7 @@ max-locals=10
 max-returns=2
 
 # Maximum number of branch for function / method body
-max-branches=2
+max-branches=3
```
Comment (extsoft) [inline .pylintrc]: @SergiyBeno sorry, but please revert this change and update the code :)

File: tests/testbed_validation_rules.py (hunk 9 above)
Comment (extsoft) [reply tests/testbed_validation_rules.py]: I don't think it is the same. At least status object has encapsulated logic on how to collect failed rules to a single string, how to calculate overall status, etc. A tuple can have these things.

File: .pylintrc (hunk 11 above)
Comment (extsoft) [inline .pylintrc]: @SergiyBeno please revert this change and redesign the code.

File: tests/testbed_validation_rules.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        return bool(re.fullmatch(self.url_pattern, grid_url))
+
+    def __str__(self) -> str:
+        return "Check Selenium grid url format"
+
+
+class CheckServerIp(Rule):
+    def is_passed(self, testbed: Testbed) -> bool:
+        server_ip = testbed.servers.server_alias.address
+        assert re.fullmatch(self.ip_pattern, server_ip)
```
Comment (extsoft) [inline tests/testbed_validation_rules.py]: @SergiyBeno  Is it better?
```suggestion
        return re.fullmatch(self.ip_pattern, server_ip)
```

File: tests/testbed_validation_rules.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    def __str__(self) -> str:
+        """Returns rule name"""
+        pass
+
+
+class CheckUserCredentials(Rule):
+    def is_passed(self, testbed: Testbed) -> bool:
+        for device in testbed:
+            device.connect()
+            assert device.connected
```
Comment (extsoft) [inline tests/testbed_validation_rules.py]: @SergiyBeno Again, why do you rely on `assert`? The method will stop working if the first device is not connected.

Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: Fixed

File: tests/testbed_validation_rules.py (hunk 12 above)
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: Fixed

File: tests/testbed_validation_rules.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+    def __str__(self) -> str:
+        return "Check server ip format"
+
+
+class Rules:
+    def __init__(self, rules: Sequence[Rule]):
+        self._rules = rules
+
+    def validate(self, testbed: Testbed) -> Tuple[bool, str]:
```
[the remark this answers is not in the document]
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: Fixed

File: tests/testbed_validation_rules.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+        for rule in self._rules:
+            try:
+                result = rule.is_passed(testbed)
+                if result:
+                    continue
+                else:
+                    return False, str(rule)
```
[the remark this answers is not in the document]
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: Fixed

File: .pylintrc (hunk 11 above)
[the remark this answers is not in the document]
Comment (SergiyBeno) [reply .pylintrc]: Reverted

Comment (SergiyBeno) [reply .pylintrc]: Reverted

File: tests/testbed_validation_rules.py (hunk 9 above)
Comment (SergiyBeno) [reply tests/testbed_validation_rules.py]: Fixed
